src/blog/understanding-ecmascript-part-4.md

```diff
@@ -0,0 +1,280 @@
+
+The spec does this by adding the following restrictions:
+
+:::ecmascript-algorithm
+> [Static Semantics: Early Errors](https://tc39.es/ecma262/#sec-grouping-operator-static-semantics-early-errors)
+>
+> `PrimaryExpression : CPEAAPL`
+>
+> It is a Syntax Error if `CPEAAPL` is not _covering_ a `ParenthesizedExpression`.
+
+:::ecmascript-algorithm
+> [Supplemental Syntax](https://tc39.es/ecma262/#sec-primary-expression)
+>
+> When processing an instance of the production
+>
+> `PrimaryExpression : CPEAAPL`
+>
+> the interpretation of the `CPEAAPL` is refined using the following grammar:
+>
+> `ParenthesizedExpression : ( Expression )`
+
+This means: if we try to use a `CPEAAPL` as a `PrimaryExpression`, it is actually an `ParenthesizedExpression` and this is its only valid production.
+
+`Expression` can never be empty, so `( )` is not a valid `ParenthesizedExpression`. Comma separated lists like `(1, 2, 3)` are created by [the comma operator](https://tc39.es/ecma262/#sec-comma-operator):
+
+```grammar
+Expression :
+AssignmentExpression
+Expression , AssignmentExpression
+```
+
+Similarly, if we try to use a `CPEAAPL` as an `ArrowParameters`, the following restrictions apply:
+
+:::ecmascript-algorithm
+> [Static Semantics: Early Errors](https://tc39.es/ecma262/#sec-arrow-function-definitions-static-semantics-early-errors)
+>
+> `ArrowParameters : CPEAAPL`
+>
+> It is a Syntax Error if `CPEAAPL` is not covering an `ArrowFormalParameters`.
+
+:::ecmascript-algorithm
+> [Supplemental Syntax](https://tc39.es/ecma262/#sec-arrow-function-definitions)
+>
+> When the production
+>
+> `ArrowParameters` : `CPEAAPL`
+>
+> is recognized the following grammar is used to refine the interpretation of `CPEAAPL`:
+>
+> `ArrowFormalParameters :`
+> `( UniqueFormalParameters )`
+
+### Other CPEAAPL restrictions
+
+There are also additional rules related to `CPEAAPL`s. For example:
+
+:::ecmascript-algorithm
+> [Static Semantics: Early Errors](https://tc39.es/ecma262/#sec-delete-operator-static-semantics-early-errors)
+>
+> `UnaryExpression: delete UnaryExpression`
+>
+> - It is a Syntax Error if the `UnaryExpression` is contained in strict mode code and the derived `UnaryExpression` is `PrimaryExpression : IdentifierReference`.
+> - It is a Syntax Error if the derived `UnaryExpression` is
+> `PrimaryExpression : CPEAAPL`
+> and `CPEAAPL` ultimately derives a phrase that, if used in place of `UnaryExpression`, would produce a  Syntax Error according to these rules. This rule is recursively applied.
+
+The first rule forbids `delete IdentifierReference` (for example, `delete foo`) in strict mode. The second rule forbids `CPEAAPL`s which would ultimately produce an `IdentifierReference`, such as `delete (foo)`, `delete ((foo))` and so on.
```

Comment:
I feel like this section doesn't add to the understanding of cover grammars, and the article would be tighter if we removed it.

Reply:
Removed it